src/state.rs

```diff
@@ -38,25 +38,65 @@ impl OPRFServer {
 }
 
 /// Advance to the next epoch on a timer
-/// This can be invoked as a background task to handle
-/// epoch advance and key rotation according to the
-/// given Config.
+/// This can be invoked as a background task to handle epoch
+/// advance and key rotation according to the given Config.
 #[instrument(skip_all)]
 pub async fn epoch_loop(state: OPRFState, config: &Config) {
     let interval =
         std::time::Duration::from_secs(config.epoch_seconds.into());
     info!("rotating epoch every {} seconds", interval.as_secs());
 
+    let starttime = time::OffsetDateTime::now_utc();
+    // Parse the epoch basetime if given.
+    let basetime: Option<time::OffsetDateTime> = config.into();
+    // If no epoch basetime was specified, use the startup time.
+    let basetime = basetime.unwrap_or(starttime);
+    assert!(
+        starttime >= basetime,
+        "epoch-basetime should be in the past"
+    );
+
+    // Calculate where we are in the epoch schedule relative to the
+    // basetime. We may need to start in the middle of the range.
+    let elapsed_epochs = (starttime - basetime) / interval;
+    let elapsed_count = elapsed_epochs.floor() as i64;
+    let epoch_count = config.last_epoch - config.first_epoch;
+    let offset = elapsed_count.rem_euclid(epoch_count.into());
+    let start = config.first_epoch + offset as u8;
+
+    // Advance to the current epoch if basetime indicates we started
+    // in the middle of a sequence.
+    if start != config.first_epoch {
+        info!(
+            "Puncturing obsolete epochs {}..{} to match basetime",
+            config.first_epoch, start
+        );
+        let mut s = state.write().expect("Failed to lock OPRFState");
+        for epoch in config.first_epoch..start {
+            s.server
+                .puncture(epoch)
+                .expect("Failed to puncture obsolete epoch");
+        }
+        s.epoch = start;
+        info!("epoch now {}", s.epoch);
+    }
+
+    // First rotation uses whatever time remains for the current epoch.
+    // This will be `interval` unless an epoch_basetime is specified.
+    let partial_seconds =
+        (1.0 - elapsed_epochs.fract()) * interval.as_secs_f64();
+    let partial = std::time::Duration::from_secs_f64(partial_seconds);
+    let mut next_rotation = starttime + partial;
```

Comment:
I had a go with this. It is cleaner than what I had before. Thanks. Kept the two variables just to split the calculation into shorter lines.

Not convinced it's correct though. Would you mind looking at the test failure and checking who's right?

Reply:
yep np, taking a look now

Reply:
couple things in the state code:

keeping both `elapsed_epochs` and `elapsed_count` in the same scope could make future modifications difficult. having a float var and an int var that is merely a conversion of the float makes it unclear which variable we should depend on for future expressions. i would recommend either collapsing it into one var and splitting the expression into two lines if need be, or redefining the variable in the second line.

the next rotation can still be shortened to:

```
let mut next_rotation = base_time + ((elapsed_count + 1) as u32 * interval);
```

i didn't know that `Mul<u64>` was not defined for `Duration`, so a cast should work in this case.

couple things in your `epoch_base_time` test:

you set `expected_time` to `config.epoch_base_time.unwrap() + advance`. since the base time is 5 seconds behind the current time, you're expecting the next rotation time to be 4 seconds in the past. setting it to `now + advance` fixes said issue.

your `advance` var is set to `config.epoch_seconds + 1`. i'm not sure why. removing the add also fixes the issue

Reply:
the year 2038 problem is for signed 32 bit integers, so it's really a year 2106 problem. also, is it likely that a user would want to set their epoch interval to a second? even if the user set the epoch interval to 5 sec, we would have approx. 680 years of epochs.

you may wonder why i'm continuing to be so nit picky with the expression i proposed. :laughing:  that's because using `ceil()` when the elapsed epoch count is already an integer (i.e. 4.0) will cause the expression to fail. also, using simple addition/arithmetic is easier to understand. using an integer to represent the elapsed epoch count also reduces the chances of error-prone code, such as what I mentioned here.

Reply:
if you're still worried about potential issues in a few hundred years you could try something like:

```
let mut next_rotation = base_time + Duration::from_secs((elapsed_count + 1) as u64 * interval.as_secs());
```

otherwise, i think the original expression i proposed should suffice for most use cases. but you never know, STAR could be a big hit when we're chilling on Mars!

Reply:
Or just use seconds since `UNIX_EPOCH` everywhere?

Anyway, I tried a compromise. Let me know what you think!

Reply:
good call with simply incrementing `next_rotation` on each iteration, clever. addresses my concerns!